aidial_adapter_bedrock/dial_api/resource.py

```diff
@@ -0,0 +1,180 @@
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(message)
+
+
+class MissingContentType(ValidationError):
+    pass
+
+
+class UnsupportedContentType(ValidationError):
+    type: str
+    supported_types: List[str]
+
+    def __init__(self, *, message: str, type: str, supported_types: List[str]):
+        self.type = type
+        self.supported_types = supported_types
+        super().__init__(message)
+
+
+class DialResource(ABC, BaseModel):
+    entity_name: str = Field(default=None)
+    supported_types: List[str] | None = Field(default=None)
+
+    @abstractmethod
+    async def download(self, storage: FileStorage | None) -> Resource: ...
+
+    @abstractmethod
+    async def guess_content_type(self) -> str | None: ...
+
+    @abstractmethod
+    async def get_resource_name(self, storage: FileStorage | None) -> str: ...
+
+    async def get_content_type(self) -> str:
+        type = await self.guess_content_type()
+
+        if not type:
+            raise MissingContentType(
+                f"Can't derive content type of the {self.entity_name}"
+            )
+
+        if (
+            self.supported_types is not None
+            and type not in self.supported_types
+        ):
+            raise UnsupportedContentType(
+                message=f"The {self.entity_name} is not one of the supported types",
+                type=type,
+                supported_types=self.supported_types,
+            )
+
+        return type
+
+
+class URLResource(DialResource):
+    url: str
+    content_type: str | None = None
+
+    @root_validator
+    def validator(cls, values):
+        values["entity_name"] = values.get("entity_name") or "URL"
+        return values
+
+    async def download(self, storage: FileStorage | None) -> Resource:
+        type = await self.get_content_type()
+        data = await _download_url(storage, self.url)
+        return Resource(type=type, data=data)
+
+    async def guess_content_type(self) -> str | None:
+        return (
+            self.content_type
+            or Resource.parse_data_url_content_type(self.url)
+            or mimetypes.guess_type(self.url)[0]
+        )
+
+    def is_data_url(self) -> bool:
+        return Resource.parse_data_url_content_type(self.url) is not None
+
+    async def get_resource_name(self, storage: FileStorage | None) -> str:
+        if self.is_data_url():
+            return f"data URL ({await self.guess_content_type()})"
+
+        name = self.url
+        if storage is not None:
+            name = await storage.get_human_readable_name(self.url)
+
+        return truncate_string(name, n=50)
+
+
+class AttachmentResource(DialResource):
+    attachment: Attachment
+
+    @validator("attachment", pre=True)
+    def parse_attachment(cls, value):
+        if isinstance(value, dict):
+            attachment = Attachment.parse_obj(value)
+            # Working around the issue of defaulting missing type to a markdown:
```

Comment:
Could you elaborate a bit more on this?

Because, if I understand correctly, defaulting to markdown should be in this part
```python
    async def guess_content_type(self) -> str | None:
        return (
            self.content_type
            or Resource.parse_data_url_content_type(self.url)
            or mimetypes.guess_type(self.url)[0] <----
        )
```
right?

Reply:
It's about working-around an unfortunate oversight in DIAL SDK itself:

https://github.com/epam/ai-dial-sdk/blob/2835107e950c89645a2b619fecba2518fa2d7bb1/aidial_sdk/chat_completion/request.py#L22

it defaults a missing `type` field to `text/markdown`, which is surprising:

```python
Attachment.parse_obj({"url": "doc.pdf"}).type == "text/markdown"
```

Reply:
I left a link to DIAL SDK in the comment.